filter.go

```diff
@@ -8,60 +8,162 @@ import (
 	manet "github.com/multiformats/go-multiaddr-net"
 )
 
+type filterEntry struct {
+	f      *net.IPNet
+	reject bool
+}
+
+// Filters is a structure representing a collection of allow/deny
+// net.IPNet filters, together with the RejectByDefault flag, which
+// represents the default filter policy.
+//
+// Note that the last policy added to the Filters is authoritative.
 type Filters struct {
+	RejectByDefault bool
+
 	mu      sync.RWMutex
-	filters map[string]*net.IPNet
+	filters []*filterEntry
 }
 
+// NewFilters constructs and returns a new set of net.IPNet filters.
+// By default, the new filter rejects no addresses.
 func NewFilters() *Filters {
 	return &Filters{
-		filters: make(map[string]*net.IPNet),
+		RejectByDefault: false,
+		filters:         make([]*filterEntry, 0),
 	}
 }
 
+func (fs *Filters) find(ff *net.IPNet) int {
+	ffs := ff.String()
+	for idx, ft := range fs.filters {
+		if ft.f.String() == ffs {
+			return idx
+		}
+	}
+
+	return -1
+}
+
+// AddDialFilter adds a reject rule to the given Filters.  Hosts
+// matching the given net.IPNet filter will be rejected, unless
+// another rule is added which states that they should be accepted.
+//
+// No effort is made to prevent duplication of filters, or to simplify
+// the filters list.
 func (fs *Filters) AddDialFilter(f *net.IPNet) {
 	fs.mu.Lock()
 	defer fs.mu.Unlock()
-	fs.filters[f.String()] = f
+
+	idx := fs.find(f)
+	if idx != -1 {
+		fs.filters[idx].reject = true
+	} else {
+		fs.filters = append(fs.filters, &filterEntry{f: f, reject: true})
+	}
+}
+
+// AddDenyFilter is an alias of AddDialFilter (which is preserved to prevent
+// an immediate breaking change.)
+func (fs *Filters) AddDenyFilter(f *net.IPNet) {
+	fs.AddDialFilter(f)
+}
+
+// AddAllowFilter adds an accept rule to the given Filters. Hosts
+// matching the given net.IPNet filter will be accepted, unless
+// another policy is added which states that they should be rejected.
+//
+// No effort is made to prevent duplication of filters, or to simplify
+// the filters list.
+func (fs *Filters) AddAllowFilter(f *net.IPNet) {
+	fs.mu.Lock()
+	defer fs.mu.Unlock()
+
+	idx := fs.find(f)
+	if idx != -1 {
+		fs.filters[idx].reject = false
+	} else {
+		fs.filters = append(fs.filters, &filterEntry{f: f, reject: false})
+	}
+}
+
+// Remove removes all net.IPNet's accept/reject rule(s) from the
+// Filters, if there are matching rules.
+//
+// Makes no distinction between whether the rule is an allow or a
+// deny.
+func (fs *Filters) Remove(ff *net.IPNet) {
```

Comment:
API ergonomics comment. This remove thing is ambiguous and the semantics are confusing. IP masking is closely related to set logic, and a `Remove` call could imply set subtraction. So if I have `10.0.0.1/8`, it's not immediately obvious what `Remove(10.1.0.0/16)` will do.

I suggest renaming it to `RemoveLiteral`, to denote that we only remove the literal range.